saleor/product/utils/variant_prices.py

```diff
@@ -117,18 +119,52 @@ def _update_or_create_listings(
             changed_variants_listings_to_update, ["discounted_price_amount"]
         )
     if changed_variant_listing_promotion_rule_to_create:
-        # after migrating to Django 4.0 we should use `update_conflicts` instead
-        # of `ignore_conflicts`
-        # https://docs.djangoproject.com/en/4.1/ref/models/querysets/#bulk-create
-        VariantChannelListingPromotionRule.objects.bulk_create(
-            changed_variant_listing_promotion_rule_to_create, ignore_conflicts=True
+        _create_variant_listing_promotion_rule(
+            changed_variant_listing_promotion_rule_to_create
         )
     if changed_variant_listing_promotion_rule_to_update:
         VariantChannelListingPromotionRule.objects.bulk_update(
             changed_variant_listing_promotion_rule_to_update, ["discount_amount"]
         )
 
 
+def _create_variant_listing_promotion_rule(variant_listing_promotion_rule_to_create):
+    with transaction.atomic():
+        rule_ids = [
+            listing.promotion_rule_id
+            for listing in variant_listing_promotion_rule_to_create
+        ]
+        listing_ids = [
+            listing.variant_channel_listing_id
+            for listing in variant_listing_promotion_rule_to_create
+        ]
+        # Lock PromotionRule and ProductVariantChannelListing before bulk_create
+        rules = PromotionRule.objects.filter(id__in=rule_ids).select_for_update()
+        variant_listings = ProductVariantChannelListing.objects.filter(
+            id__in=listing_ids
+        ).select_for_update()
+        # Do not create VariantChannelListingPromotionRule for rules that were deleted.
+        if len(rules) < len(rule_ids):
+            variant_listing_promotion_rule_to_create = [
+                listing
+                for listing in variant_listing_promotion_rule_to_create
+                if listing.promotion_rule_id in [rule.id for rule in rules]
+            ]
+        if len(variant_listings) < len(listing_ids):
+            variant_listing_promotion_rule_to_create = [
+                listing
+                for listing in variant_listing_promotion_rule_to_create
+                if listing.variant_channel_listing_id
+                in [listing.id for listing in variant_listings]
+            ]
```

Comment:
```suggestion
rule_id_set = {rule.id for rule in rules}
if len(rules) < len(rule_ids):
    variant_listing_promotion_rule_to_create = [
        listing
        for listing in variant_listing_promotion_rule_to_create
        if listing.promotion_rule_id in rule_id_set
    ]

listing_id_set = {listing.id for listing in variant_listings}
if len(variant_listings) < len(listing_ids):
    variant_listing_promotion_rule_to_create = [
        listing
        for listing in variant_listing_promotion_rule_to_create
        if listing.variant_channel_listing_id in listing_id_set
    ]
```

Just a hint :)

In this updated code, rule_id_set and listing_id_set are sets containing the IDs from rules and variant_listings, respectively. The membership tests are then performed using these sets, which should be faster than using lists, especially if the number of rules or listings is large.